So I add manually 
`args.add("-Djava.library.path=.");`
but I continue to have the following error. Even if the native image compilation is done with jni flag.
Any idea of methodology to find my mistake ? Because, i don't see any error...


Caused by: java.lang.UnsatisfiedLinkError: org.acme.JniWrapper.getString()Ljava/lang/String; [symbol: Java_org_acme_JniWrapper_getString or Java_org_acme_JniWrapper_getString__]
	at com.oracle.svm.jni.access.JNINativeLinkage.getOrFindEntryPoint(JNINativeLinkage.java:145)

--- a/core/creator/src/main/java/io/quarkus/creator/phase/nativeimage/NativeImagePhase.java
+++ b/core/creator/src/main/java/io/quarkus/creator/phase/nativeimage/NativeImagePhase.java
@@ -381,15 +381,23 @@ public class NativeImagePhase implements AppCreationPhase<NativeImagePhase>, Nat
                     }
                 }
 
-                enableSslNative = properties.getProperty("quarkus.ssl.native") != null
-                        ? Boolean.parseBoolean(properties.getProperty("quarkus.ssl.native"))
+                final String enableSslNativeFromProperties = properties.getProperty("quarkus.ssl.native");
+                enableSslNative = enableSslNativeFromProperties != null
+                        ? Boolean.parseBoolean(enableSslNativeFromProperties)
                         : false;
+                if (!enableJni) {
+                    final String enableJniFromProperties = properties.getProperty("quarkus.jni.enable");
+                    enableJni = enableJniFromProperties != null
+                            ? Boolean.parseBoolean(enableJniFromProperties)
+                            : false;
+                }
             }
             if (enableSslNative) {
                 enableHttpsUrlHandler = true;
                 enableJni = true;
                 enableAllSecurityServices = true;
             }
+
             if (additionalBuildArgs != null) {
                 command.addAll(additionalBuildArgs);
             }

--- a/core/deployment/src/main/java/io/quarkus/deployment/JniProcessor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/JniProcessor.java
@@ -1,0 +1,37 @@
+package io.quarkus.deployment;
+
+import java.util.List;
+
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.JniBuildItem;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+public class JniProcessor {
+
+    JniConfig jni;
+
+    @ConfigRoot(phase = ConfigPhase.BUILD_TIME)
+    static class JniConfig {
+        /**
+         * Paths of library to load.
+         */
+        @ConfigItem
+        List<String> libraryPaths;
+
+        /**
+         * Enable JNI support.
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean enable = false;
+    }
+
+    @BuildStep
+    void setupJni(BuildProducer<JniBuildItem> jniProducer) {
+        if ((jni.enable) || !jni.libraryPaths.isEmpty()) {
+            jniProducer.produce(new JniBuildItem(jni.libraryPaths));
+        }
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java
@@ -1,0 +1,24 @@
+package io.quarkus.deployment.builditem;
+
+import java.util.Collections;
+import java.util.List;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class JniBuildItem extends MultiBuildItem {
+
+    private final List<String> libraryPaths;
+
+    public JniBuildItem() {
+        this(Collections.emptyList());
+    }
+
+    public JniBuildItem(List<String> libraryPaths) {
+        this.libraryPaths = libraryPaths;
+    }
+
+    public List<String> getLibraryPaths() {
+        return libraryPaths;
+    }
+
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java
@@ -3,6 +3,7 @@ package io.quarkus.deployment.steps;
 import static io.quarkus.gizmo.MethodDescriptor.ofConstructor;
 import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
 
+import java.io.File;
 import java.lang.reflect.Modifier;
 import java.util.List;
 import java.util.Optional;
@@ -125,32 +126,22 @@ class MainClassBuildStep {
 
         // Set the SSL system properties
         if (!javaLibraryPathAdditionalPaths.isEmpty()) {
-            // FIXME: this is the code we should use but I couldn't get GraalVM to work with a java.library.path containing multiple paths.
-            // We need to dig further but for now, we need this to work.
-            // ResultHandle javaLibraryPath = mv.newInstance(ofConstructor(StringBuilder.class, String.class),
-            //         mv.invokeStaticMethod(ofMethod(System.class, "getProperty", String.class, String.class), mv.load(JAVA_LIBRARY_PATH)));
-            // for (JavaLibraryPathAdditionalPathBuildItem javaLibraryPathAdditionalPath : javaLibraryPathAdditionalPaths) {
-            //     ResultHandle javaLibraryPathLength = mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "length", int.class), javaLibraryPath);
-            //     mv.ifNonZero(javaLibraryPathLength).trueBranch()
-            //             .invokeVirtualMethod(ofMethod(StringBuilder.class, "append", StringBuilder.class, String.class), javaLibraryPath, mv.load(File.pathSeparator));
-            //     mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "append", StringBuilder.class, String.class), javaLibraryPath,
-            //             mv.load(javaLibraryPathAdditionalPath.getPath()));
-            // }
-            // mv.invokeStaticMethod(ofMethod(System.class, "setProperty", String.class, String.class, String.class),
-            //         mv.load(JAVA_LIBRARY_PATH), mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "toString", String.class), javaLibraryPath));
-
-            ResultHandle isJavaLibraryPathEmpty = mv.invokeVirtualMethod(ofMethod(String.class, "isEmpty", boolean.class),
+            ResultHandle javaLibraryPath = mv.newInstance(ofConstructor(StringBuilder.class, String.class),
                     mv.invokeStaticMethod(ofMethod(System.class, "getProperty", String.class, String.class),
                             mv.load(JAVA_LIBRARY_PATH)));
-
-            BytecodeCreator inGraalVMCode = mv
-                    .ifNonZero(mv.invokeStaticMethod(ofMethod(ImageInfo.class, "inImageRuntimeCode", boolean.class)))
-                    .trueBranch();
-
-            inGraalVMCode.ifNonZero(isJavaLibraryPathEmpty).trueBranch().invokeStaticMethod(
-                    ofMethod(System.class, "setProperty", String.class, String.class, String.class),
-                    inGraalVMCode.load(JAVA_LIBRARY_PATH),
-                    inGraalVMCode.load(javaLibraryPathAdditionalPaths.iterator().next().getPath()));
+            for (JavaLibraryPathAdditionalPathBuildItem javaLibraryPathAdditionalPath : javaLibraryPathAdditionalPaths) {
+                ResultHandle javaLibraryPathLength = mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "length", int.class),
+                        javaLibraryPath);
+                mv.ifNonZero(javaLibraryPathLength).trueBranch()
+                        .invokeVirtualMethod(ofMethod(StringBuilder.class, "append", StringBuilder.class, String.class),
+                                javaLibraryPath, mv.load(File.pathSeparator));
+                mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "append", StringBuilder.class, String.class),
+                        javaLibraryPath,
+                        mv.load(javaLibraryPathAdditionalPath.getPath()));
+            }
+            mv.invokeStaticMethod(ofMethod(System.class, "setProperty", String.class, String.class, String.class),
+                    mv.load(JAVA_LIBRARY_PATH),
+                    mv.invokeVirtualMethod(ofMethod(StringBuilder.class, "toString", String.class), javaLibraryPath));
         }
 
         if (sslTrustStoreSystemProperty.isPresent()) {

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
@@ -16,6 +16,7 @@ import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
 import io.quarkus.deployment.builditem.JavaLibraryPathAdditionalPathBuildItem;
+import io.quarkus.deployment.builditem.JniBuildItem;
 import io.quarkus.deployment.builditem.SslNativeConfigBuildItem;
 import io.quarkus.deployment.builditem.SslTrustStoreSystemPropertyBuildItem;
 import io.quarkus.deployment.builditem.SystemPropertyBuildItem;
@@ -39,6 +40,7 @@ class SubstrateConfigBuildStep {
     void build(SslContextConfigurationTemplate sslContextConfigurationTemplate,
             List<SubstrateConfigBuildItem> substrateConfigBuildItems,
             SslNativeConfigBuildItem sslNativeConfig,
+            List<JniBuildItem> jniBuildItems,
             List<ExtensionSslNativeSupportBuildItem> extensionSslNativeSupport,
             BuildProducer<SubstrateProxyDefinitionBuildItem> proxy,
             BuildProducer<SubstrateResourceBundleBuildItem> resourceBundle,
@@ -87,8 +89,7 @@ class SubstrateConfigBuildStep {
                 Path linuxPath = linuxLibDirectory.resolve(LIB_SUN_EC);
 
                 // We add . as it might be useful in a containerized world
-                // FIXME: it seems GraalVM does not support having multiple paths in java.library.path
-                //javaLibraryPathAdditionalPath.produce(new JavaLibraryPathAdditionalPathBuildItem("."));
+                javaLibraryPathAdditionalPath.produce(new JavaLibraryPathAdditionalPathBuildItem("."));
                 if (Files.exists(linuxPath)) {
                     // On Linux, the SunEC library is in jre/lib/amd64/
                     // This is useful for testing or if you have a similar environment in production
@@ -112,8 +113,19 @@ class SubstrateConfigBuildStep {
                 }
             }
         }
-
         nativeImage.produce(new SubstrateSystemPropertyBuildItem("quarkus.ssl.native", sslNativeEnabled.toString()));
+
+        if (!jniBuildItems.isEmpty()) {
+            for (JniBuildItem jniBuildItem : jniBuildItems) {
+                if (jniBuildItem.getLibraryPaths() != null && !jniBuildItem.getLibraryPaths().isEmpty()) {
+                    for (String path : jniBuildItem.getLibraryPaths()) {
+                        javaLibraryPathAdditionalPath
+                                .produce(new JavaLibraryPathAdditionalPathBuildItem(path));
+                    }
+                }
+            }
+            nativeImage.produce(new SubstrateSystemPropertyBuildItem("quarkus.jni.enable", "true"));
+        }
     }
 
     private Boolean isSslNativeEnabled(SslNativeConfigBuildItem sslNativeConfig,

